Stream aggregation: reduce the default staleness interval (#11105)
VictoriaMetrics/VictoriaMetrics: docs/victoriametrics/stream-aggregation/README.md

If counter-specific outputs, such as `total*`, `rate*`, and `increase*`, produce values that are significantly higher than anticipated, then check the `vm_streamaggr_counter_resets_total` metric. This metric increments each time when [counter reset event](https://docs.victoriametrics.com/victoriametrics/keyconcepts/#counter) happens and could be caused by duplication or collision of raw samples. If you observe duplication or collision, try solving this problem by either fixing the source of these metrics or by [deduplicating](https://docs.victoriametrics.com/victoriametrics/stream-aggregation/#deduplication) these samples before aggregation.

> ## Data delay and staleness

Reviewer: This change will break backward compatibility in the docs. It is better to keep old anchor names and add new section/subsection.
Or keep the old anchor id via the following construction:
```
## Data delay and staleness {#staleness}
```

Author: I tried using two sections, but some content was repeated between them. I prefer to keep data delay and staleness in a single section to cover the case more clearly, so I chose `## Data delay and staleness {#staleness}`.

> Since stream aggregation processes input samples in a streaming manner and flushes results when a specific aggregation window closes, aggregation results can be heavily affected by data delay(it can be observed via the `vm_streamaggr_samples_lag_seconds_bucket` metric).

Reviewer: ```suggestion
Stream aggregation processes input samples in a streaming manner and flushes results once per specified `interval`. Because of this, aggregation results can be heavily affected by data delays (see `vm_streamaggr_samples_lag_seconds_bucket` metric).
```

In particular:

> 1. If input samples are delayed for multiple intervals, stream aggregation will not see any input samples during that period and will not generate any results for that time, which causes gaps in the output.

Reviewer: ```suggestion
1. Stream aggregation won't produce results if input samples are delayed for multiple aggregation intervals, causing gaps in the output.
```

> 2. If delayed samples are delivered later, the aggregation result for the next aggregation window can be affected by those delayed samples, depending on their arrival order and how long the delay is compared to the interval.

Reviewer: ```suggestion
2. Delayed and out-of-order samples can inflate or skew correctness of aggregation resuts.
```

If you prefer consistency in aggregation results and do not want delayed data to affect the next aggregation window, drop all potentially delayed samples via [ignore_old_samples](https://docs.victoriametrics.com/victoriametrics/stream-aggregation/#ignoring-old-samples).

> If you prefer to have the accumulated changes from delayed data reflected in aggregation windows after the delay, increase `staleness_interval` in the [stream aggregation config](https://docs.victoriametrics.com/victoriametrics/stream-aggregation/configuration/#stream-aggregation-config).

Reviewer: ```suggestion
Dropping delayed and out-of-order samples could result into data gaps in produced results. Keeping delayed and out-of-order samples may inflate produced results. It is up to user to decide how to set aggregation for such cases:

- If you prefer consistency in aggregation results and do not want delayed data to affect the next aggregation window, drop all potentially delayed samples via [ignore_old_samples](https://docs.victoriametrics.com/victoriametrics/stream-aggregation/#ignoring-old-samples).
- If you prefer to have the accumulated changes from delayed data to be counted in aggregation windows after the delay, increase staleness_interval in the [stream aggregation config](https://docs.victoriametrics.com/victoriametrics/stream-aggregation/configuration/#stream-aggregation-config).
```

Author: >Dropping delayed and out-of-order samples could result into data gaps in produced results.

This feels a bit misleading, as it suggests that keeping delayed samples will remove the gap in the results.

>Keeping delayed and out-of-order samples

User can't keep out-of-order samples.

Rephrased in https://github.com/VictoriaMetrics/VictoriaMetrics/pull/11105/changes/a849df3d02dd65b6ec2ec1b78798d9325f075987.

- By increasing the `interval` option at [stream aggregation config](https://docs.victoriametrics.com/victoriametrics/stream-aggregation/configuration/#stream-aggregation-config), so it covers the expected
  delays in data ingestion pipelines.
- By specifying the `staleness_interval` option at [stream aggregation config](https://docs.victoriametrics.com/victoriametrics/stream-aggregation/configuration/#stream-aggregation-config), so it covers the expected

>   delays in data ingestion pipelines. By default, the `staleness_interval` is equal to `interval`.

Reviewer: Isn't this recommendation contradicts with the code change? To avoid spikes, this PR reduces the staleness interval. And this part of doc recommends increasing staleness interval to reduce spikes.

Author: The doc is correct. Staleness has opposite effects on `increase` and `increase_prometheus`: when a series is marked as stale, `increase` can produce spikes, while `increase_prometheus` can produce unexpectedly lower results if the user expects accumulated changes to be reflected. So the logic in the doc is correct, but it can be tricky to understand.

Author: Tried to improve doc in https://github.com/VictoriaMetrics/VictoriaMetrics/pull/11105/changes/5b39a9b738cbe38c8ba30243a9c5c3f468976ef1, please check.
